src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpMetricSerializer.cs

```diff
@@ -0,0 +1,352 @@
+
+                    foreach (var metricPoint in metric.GetMetricPoints())
+                    {
+                        var lastValue = metricPoint.GetGaugeLastValueDouble();
+                        writePosition = WriteNumberDataPoint(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Gauge_Data_Points, metricPoint, lastValue);
+                    }
+
+                    ProtobufSerializer.WriteReservedLength(buffer, metricTypeLengthPosition, writePosition - (metricTypeLengthPosition + ReserveSizeForLength));
+                    break;
+                }
+
+            case MetricType.Histogram:
+                {
+                    writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Metric_Data_Histogram, ProtobufWireType.LEN);
+                    int metricTypeLengthPosition = writePosition;
+                    writePosition += ReserveSizeForLength;
+
+                    // TODO: DO WORK HERE
+
+                    /*message Histogram {
+                          repeated HistogramDataPoint data_points = 1;
+                          AggregationTemporality aggregation_temporality = 2;
+                        }
+                    */
+
+                    ProtobufSerializer.WriteReservedLength(buffer, metricTypeLengthPosition, writePosition - (metricTypeLengthPosition + ReserveSizeForLength));
+                    break;
+                }
+
+            case MetricType.ExponentialHistogram:
+                {
+                    writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.Metric_Data_Exponential_Histogram, ProtobufWireType.LEN);
+                    int metricTypeLengthPosition = writePosition;
+                    writePosition += ReserveSizeForLength;
+
+                    // TODO: DO WORK HERE
+
+                    /*message ExponentialHistogram {
+                          repeated ExponentialHistogramDataPoint data_points = 1;
+                          AggregationTemporality aggregation_temporality = 2;
+                        }
+                    */
+
+                    ProtobufSerializer.WriteReservedLength(buffer, metricTypeLengthPosition, writePosition - (metricTypeLengthPosition + ReserveSizeForLength));
+                    break;
+                }
+        }
+
+        ProtobufSerializer.WriteReservedLength(buffer, metricLengthPosition, writePosition - (metricLengthPosition + ReserveSizeForLength));
+        return writePosition;
+    }
+
+    private static int WriteNumberDataPoint<T>(byte[] buffer, int writePosition, int fieldNumber, in MetricPoint metricPoint, T value)
+    {
+        writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, fieldNumber, ProtobufWireType.LEN);
+        int dataPointLengthPosition = writePosition;
+        writePosition += ReserveSizeForLength;
+
+        if (typeof(T) == typeof(long))
```

Comment:
As you know if you are calling from long / double block. can we have an overload instead of this check here?

Reply:
I personally don't mind it. This type check should be elided by the JIT since `T` is a value type. Specialized code will be generated and things which resolve to always false should be removed.

Reply:
Example: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACAZQAtsoAHAGW2ADoASgFcAdhgCW+GAG4axAMwMyDAMIMA3jQbali4igYBZABQBKDVp1WAKjFwZjdLqdnUrNuw7oARF5e0Avv4MwQDaAFLiGADiMCIwUOJgxhgAnhwwEABmxgA2ECIA5qamaAyRMXEJSSnpmTkAJhBCwLkwJQC6wQpKBrb2ADzWAHzG1gwAbti5Qu3Bmm7u2uJZDLUZ2WPmALzbDGkbOflFJcFWC0tLxOQAnMZ5BcXGEMAAVjBgGKZTMzAAhH5FksgkD3DBcrgYAwVmsDvUtgxdvs6psmi02qdQeczpdrndjGjWu1nm8Pl8frMAa5LgwQVYQQEgA==

Notice the specialized implementations generated (`C.Test[[System.Int64, System.Private.CoreLib]](Int64)` and `C.Test[[System.Double, System.Private.CoreLib]](Double)`). The type check and casts get elided.